File: day4/part1.rb
data = File.read("data.txt")

passport_hashes = data.split("\n\n").map do |passport_string|
  passport_string.split.inject({}) do |passport_hash, field_string|
    key, val = field_string.split(":")
    passport_hash[key] = val
    passport_hash
  end
end

puts passport_hashes

RequiredKeys = %w[ byr iyr eyr hgt hcl ecl pid ]

number_of_passports_with_required_fields =
  passport_hashes.inject(0) do |n, passport|
    n += 1 if (RequiredKeys - passport.keys).empty?
    n
  end

puts number_of_passports_with_required_fields

File: day4/part1.hs
module Main where
import Data.List ((\\))

type Passport = [PassportField]
type PassportField = (String, String)

main = do
    inputString <- readFile "data.txt"
    let passports = parsePassports inputString
    print passports
    let passportCount = length $ filter hasRequiredFields passports 
    print $ show passportCount



-- Testing a passport for required keys

hasRequiredFields :: Passport -> Bool
hasRequiredFields passport = null $ requiredKeys \\ keys passport

requiredKeys = ["byr","iyr","eyr","hgt","hcl","ecl","pid"]

keys :: Passport -> [String]
keys passport = ks 
    where (ks, _) = unzip passport



-- Parsing a list of Passports from a string

parsePassports :: String -> [Passport]
parsePassports s = map parsePassport (splitOnEmptyLine s)

parsePassport :: String -> Passport
parsePassport s = map parseField (words s)

parseField :: String -> PassportField
parseField s = (key, value)
    where 
      (key,_:value) = break (==':') s

splitOnEmptyLine :: String -> [String]
splitOnEmptyLine s = firstElem : restOfTheElems
    where firstElem = unlines firstItemLines
          (firstItemLines, remainingLines) = break null (lines s)
          restOfTheElems
            | null remainingLines = []
            | otherwise = splitOnEmptyLine $ unlines $ tail remainingLines

